### PR TITLE
Add documentation for indexing.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [X] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

It adds documentation for DataNode indexing.

#### Additional information
- Affected modules and functionalities:

Documentation

- Key points relevant for the review:

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [X] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [X] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2219

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -38,8 +38,7 @@ NVIDIA DALI Documentation
 
    pipeline
    data_types
-   supported_ops
-   math
+   data_manip
```
Comment (JanuszL) [inline docs/index.rst:41]:
```suggestion
   data_manipulation
```
I don't think we need to shorten it here.

File: docs/index.rst [hunk 2]
```diff
@@ -38,8 +38,7 @@ NVIDIA DALI Documentation
 
    pipeline
    data_types
-   supported_ops
```
Comment (JanuszL) [inline docs/index.rst:41]: Should we move supported_ops into data manipulation?

File: docs/indexing.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+The simplest case is scalar indexing with a constant index::
+
+    raw_files = fn.readers.file(...)
+    third_bytes = raw_files[2]
+
+The snippet above extracts the third (index 2) byte from the files.
+
+.. note::
+    The batch dimension is implicit and cannot be indexed. In this example, the index 2
+    is broadcast to the whole batch. The section :ref:`Run-time indices` for per-sample indexing.
```
Comment (JanuszL) [inline docs/indexing.rst:20]:
```suggestion
    is broadcast to the whole batch. Check the section :ref:`Run-time indices` for per-sample indexing.
```

File: docs/indexing.rst [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+DALI data nodes can be indexed and sliced using familiar Python syntax for array indexing.
+
+Scalar indexing
+~~~~~~~~~~~~~~~
+
+The simplest case is scalar indexing with a constant index::
+
+    raw_files = fn.readers.file(...)
+    third_bytes = raw_files[2]
```
Comment (JanuszL) [inline docs/indexing.rst:14]: Maybe it is better to show the example with the shape in the tensor (and how easy is to extract w and h now) - is it more real life.

File: docs/indexing.rst [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+The index can contain multiple axes, separated by comma. If an index is a scalar, the corresponding
+dimension is removed from the output::
+
+    image = fn.decoders.image(jpegs, device="mixed")  # RGB images in HWC layout
+    red =   image[:,:,0]
+    green = image[:,:,1]
+    blue =  image[:,:,2]
+
+The ``red``, ``green``, ``blue`` are 2D tensors.
```
Comment (JanuszL) [inline docs/indexing.rst:72]: So ```
    red =   image[:,:,:1]
    green = image[:,:,1:2]
    blue =  image[:,:,2:3]
```
will keep the dimensions, right?

File: docs/indexing.rst [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    video = ... # a 4D tensor in FHWC layout
+    fifth_frame = video[5]         # equivalent to...
+    fifth_frame = video[5,:]       # ...equivalent to...
+    fifth_frame = video[5,:,:]     # ...equivalent to...
+    fifth_frame = video[5,:,:,:]
+
+Strided slices
+~~~~~~~~~~~~~~
+
+Indexing with a custom step is not implemented.
```
Comment (JanuszL) [inline docs/indexing.rst:93]:
```suggestion
Indexing with a custom step is not supported.
```

File: docs/indexing.rst [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    blue =  image[:,:,2]
+
+The ``red``, ``green``, ``blue`` are 2D tensors.
+
+Using ranges doesn't remove dimensions::
+
+    letterboxed = image[20:-20,:,:]   # the result is still a 3D HWC tensor
+
+.. note::
+    See also :func:`fn.crop` and :func:`fn.slice` for operations tailored for image processing.
```
Comment (JanuszL) [inline docs/indexing.rst:79]: Do we have any perf numbers comparing the indexing and slice?

File: docs/indexing.rst [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+DALI tensors can have a layout specifier which affects how the data is interpreted by some
+operators - typically, an image would have ``"HWC"`` layout.
+
+When applying a scalar index to an axis, that axis is removed from the output along with the
+layout name for this axis::
+
+    image = ...             # layout HWC
+    first_row = image[0]    # layout WC
+    last_col = image[:,-1]  # lauout HWC
```
Comment (JanuszL) [inline docs/indexing.rst:115]:
```suggestion
    last_col = image[:,-1]  # layout HWC
```

File: docs/indexing.rst (hunk 5 above)
Comment (mzient) [reply docs/indexing.rst:72]: Yes.

File: docs/indexing.rst (hunk 6 above)
Comment (mzient) [reply docs/indexing.rst:93]: I'd prefer to call it "not implemented", because we intend to support it.

File: docs/indexing.rst (hunk 3 above)
Comment (mzient) [reply docs/indexing.rst:20]: Hmm... what happened to my `see`?

File: docs/index.rst (hunk 2 above)
Comment (mzient) [reply docs/index.rst:41]: That's what I'd like to discuss.

File: docs/indexing.rst (hunk 6 above)
Comment (JanuszL) [reply docs/indexing.rst:93]: I think `implemented` is better for the code documentation, supported sounds more like a term from manual.

File: docs/indexing.rst (hunk 3 above)
Comment (JanuszL) [reply docs/indexing.rst:20]: 🤷

File: docs/indexing.rst (hunk 4 above)
Comment (klecki) [reply docs/indexing.rst:14]: This sounds like a good idea.

File: docs/indexing.rst [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+Run-time indices
+~~~~~~~~~~~~~~~~
+
+Indexing by constant is often insufficient. With DALI, you can use a result of other computations
+to access tensor elements::
+
+    length = fn.shapes(raw_files)[0]
+    index = fn.cast(fn.random.uniform(range=(0, 1)) * length, dtype=dali.types.INT64)
+    random_byte = raw_files[index]
```
Comment (klecki) [inline docs/indexing.rst:40]: Won't we get values in range [0, length] due to rounding, thus doing and out of bounds access?

Comment (klecki) [reply docs/indexing.rst:40]: I also think we should focus on this line more, maybe first constructing it by hand or something? This random is a bit cryptic, and also there is this length already using subscript.

File: docs/indexing.rst [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    header = raw_files[:16]
+
+If the start of the slice is omitted, the slice starts at 0. If the end is omitted, the slice
+ends at the end of given axis. Negative indices can be used for both start and end of the slice.
+Either end can be a constant, a run-time value (a DataNode) or can be skipped.
+
+Multidimensional indexing
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The index can contain multiple axes, separated by comma. If an index is a scalar, the corresponding
```
Comment (klecki) [inline docs/indexing.rst:64]: I'm thinking how easy one can confuse scalars and constants here. Maybe we can find some better wording? Even writing `non-range` might help.

Also, the "dimension removal" can be bit misleading imo, maybe we should add a sentence, that the data is there, we just don't keep the 1-element-sized dummy axis?

File: docs/indexing.rst [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+dimension is removed from the output::
+
+    image = fn.decoders.image(jpegs, device="mixed")  # RGB images in HWC layout
+    red =   image[:,:,0]
+    green = image[:,:,1]
+    blue =  image[:,:,2]
```
Comment (klecki) [inline docs/indexing.rst:70]: General comment, I think we should add spaces after commas, as it's hard to sometimes separate dimensions visually, as the colons and commas stick together, especially if there are few of them side by side:
```suggestion
    red =   image[:, :, 0]
    green = image[:, :, 1]
    blue =  image[:, :, 2]
```

File: docs/indexing.rst [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+The ``red``, ``green``, ``blue`` are 2D tensors.
+
+Using ranges doesn't remove dimensions::
+
+    letterboxed = image[20:-20,:,:]   # the result is still a 3D HWC tensor
```
Comment (klecki) [inline docs/indexing.rst:76]: Maybe it's worth to compare it to R, G, B slices but with the dummy dimension so:
```
    red =   image[:, :, 0:1]
    green = image[:, :, 1:2]
    blue =  image[:, :, 2:3]
```

File: docs/indexing.rst (hunk 6 above)
Comment (klecki) [reply docs/indexing.rst:93]: not supported yet?

File: docs/indexing.rst [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+size 1 in the output::
+
+    trailing_channel = grayscale[:,:,dali.newaxis]
+    leading_channel = grayscale[dali.newaxis]
+
+Layout specifiers
+~~~~~~~~~~~~~~~~~
+
+DALI tensors can have a layout specifier which affects how the data is interpreted by some
+operators - typically, an image would have ``"HWC"`` layout.
```
Comment (klecki) [inline docs/indexing.rst:108]: Could you maybe add a link to: https://docs.nvidia.com/deeplearning/dali/user-guide/docs/data_types.html#data-layouts

File: docs/indexing.rst (hunk 8 above)
Comment (klecki) [reply docs/indexing.rst:115]:
```suggestion
    last_col = image[:,-1]  # layout HC
```

File: docs/index.rst (hunk 1 above)
Comment (klecki) [reply docs/index.rst:41]: How about something like:
```suggestion
.. toctree::
   :hidden:
   :caption: Data Manipulation
   graph_building # Put some links to Pipeline doc and Getting started here
   indexing
   math
   supported_ops
```

Comment (klecki) [reply docs/index.rst:41]: To keep it top-level

File: docs/indexing.rst [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+Indexing by constant is often insufficient. With DALI, you can use a result of other computations
+to access tensor elements::
+
+    raw_files = fn.readers.file(...)
+    length = fn.shapes(raw_files)[0]
+
+    # calculate a random index from 0 to file_length-1
+    random_01 = fn.random.uniform(range=(0, 1))  # random numbers in range [0..1)
+    index = fn.floor(random_01 * length)  # calculate indices from [0..length)
```
Comment (JanuszL) [inline docs/indexing.rst:48]: Or use cast?

File: docs/indexing.rst (hunk 10 above)
Comment (JanuszL) [reply docs/indexing.rst:76]: On the other hand that is how python works as well.

**Review by JanuszL — APPROVED** [review]

File: docs/indexing.rst (hunk 14 above)
Comment (klecki) [reply docs/indexing.rst:48]: It rounds.

File: docs/indexing.rst (hunk 3 above)
Comment (mzient) [reply docs/indexing.rst]: Fixed.

File: docs/indexing.rst (hunk 7 above)
Comment (mzient) [reply docs/indexing.rst:91]: I don't think there's any reason for a difference there - they use the same kernel. Indexing might be ever-so-slightly more efficient when collapsing is used, but I'm not pushing it too much.

File: docs/indexing.rst (hunk 9 above)
Comment (mzient) [reply docs/indexing.rst:52]: Done.

File: docs/indexing.rst (hunk 7 above)
Comment (JanuszL) [reply docs/indexing.rst:91]: Understood

File: docs/indexing.rst (hunk 10 above)
Comment (mzient) [reply docs/indexing.rst:76]: @klecki We don't keep the axis. That''s the point.

Comment (mzient) [reply docs/indexing.rst:76]: I'll reword it to "indexing" (scalar) and "slicing" (range).

File: docs/indexing.rst (hunk 6 above)
Comment (mzient) [reply docs/indexing.rst:105]: I still think that "not implemented" has a slightly different vibe.
For me, "not implemented" means that it's a "simple matter of programming" to add the feature and that we're aware of the incompleteness of the existing solution.
Conversely, "not supported" sounds like we're not just supporting it. "not supported **yet**" is more in the "nice to have" realm.

Comment (JanuszL) [reply docs/indexing.rst:105]: > For me, "not implemented" means that it's a "simple matter of programming" to add the feature and that we're aware of the incompleteness of the existing solution.

I don't think the user care why it doesn't work now and how difficult is it to add. All it matters is "we're not just supporting it.".

File: docs/indexing.rst [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+Negative indices can be used to index the tensor starting from the end. The index of -1 denotes
+the last element::
+
+    channels = sizes[-1]   # channels go last
+    widths = sizes[-2]     # widths are the innermost dimension after channels
+
+Indexing with run-time values
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Indexing by constant is often insufficient. With DALI, you can use a result of other computations
```
Comment (awolant) [inline docs/indexing.rst:45]: Maybe `by a constant`. But I'm not sure, tbh.

File: docs/indexing.rst [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    header = raw_files[:16]  # extract 16-byte headers from files in the batch
+
+If the start of the slice is omitted, the slice starts at 0. If the end is omitted, the slice
+ends at the end of given axis. Negative indices can be used for both start and end of the slice.
+Either end can be a constant, a run-time value (a DataNode) or can be skipped.
+
+Multidimensional selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The slection can contain multiple axes, separated by comma. If an index is a scalar, the
```
Comment (awolant) [inline docs/indexing.rst:82]: Again, maybe `by a comma` ?

File: docs/indexing.rst [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+When indexing and slicing multidimensional data, the trailing dimenions can be omitted. This is
+equivalent to passing a full-range slice to all trailing dimensions::
+
+    wide = letterboxed[20:-20,:,:]   # slice height, keep width and channels
+    wide = letterboxed[20:-20,:]     # this line is equivalent to the previous one
+    wide = letterboxed[20:-20]       # ...and so is this one
+
+.. note::
+    See also :func:`fn.crop` and :func:`fn.slice` for operations tailored for image processing.
```
Comment (awolant) [inline docs/indexing.rst:104]: Maybe it would be useful to add very small code example on how you can replace slice with indexing somewhere in this doc?

**Review by awolant — APPROVED** [review]

File: docs/indexing.rst (hunk 16 above)
Comment (mzient) [reply docs/indexing.rst:82]: True.

Comment (mzient) [reply docs/indexing.rst:82]: or rather commas?

**Review by JanuszL — APPROVED** [review]
